graspologic/models/edge_swaps.py

```diff
@@ -0,0 +1,213 @@
+        """
+
+        if isinstance(self.adjacency, np.ndarray):
+            adj_length = len(self.adjacency)
+
+        else:
+            adj_length = self.adjacency.get_shape()[0]
+        edge_list = []
+
+        for i in range(adj_length):
+            for j in range(i, adj_length):
+                if self.adjacency[i, j] == 1:
+                    edge_list.append([i, j])
+        return edge_list
+
+    @staticmethod
+    @nb.jit
+    def _edge_swap(
+        adjacency: Union[np.ndarray, csr_matrix], edge_list: List
+    ) -> Tuple[Union[np.ndarray, csr_matrix], List]:
+        """
+        Performs the edge swap on the adjacency matrix. If adjacency is
+        np.ndarray, then nopython=True is used in numba, but if adjacency
+        is csr_matrix, then forceobj=True is used in numba
+
+        Parameters
+        ----------
+        adjacency : np.ndarray OR csr_matrix, shape (n_verts, n_verts)
+            The initial adjacency matrix in which edge swaps are performed on it
+
+        edge_list : np.ndarray, shape (n_verts, 2)
+            The corresponding edge_list of adjacency
+
+        seed : int (default 1234), optional
+            The seed with which we seed the process of choosing two random edges
+            from the graph
+
+        Returns
+        -------
+        adjacency : np.ndarray OR csr_matrix, shape (n_verts, n_verts)
+            The adjancency matrix after an edge swap is performed on the graph
+
+        edge_list : np.ndarray (n_verts, 2)
+            The edge_list after an edge swap is perfomed on the graph
+        """
+
+        # choose two indices at random
+        orig_inds = np.random.choice(len(edge_list), size=2, replace=False)
+
+        u, v = edge_list[orig_inds[0]]
+
+        # two types of swap orientations for undirected graph
+        if np.random.rand() < 0.5:
+            x, y = edge_list[orig_inds[1]]
+        else:
+            y, x = edge_list[orig_inds[1]]
+
+        # ensures no initial loops
+        if u == v or x == y:
+            return adjacency, edge_list
+
+        # ensures no loops after swap (must be swap on 4 distinct nodes)
+        if u == x or v == y:
+            return adjacency, edge_list
+
+        # save edge values
+        w_uv = adjacency[u, v]
+        w_xy = adjacency[x, y]
+        w_ux = adjacency[u, x]
+        w_vy = adjacency[v, y]
+
+        # ensures no initial multigraphs
+        if w_uv > 1 or w_xy > 1:
+            return adjacency, edge_list
+
+        # ensures no multigraphs after swap
+        if w_ux >= 1 or w_vy >= 1:
+            return adjacency, edge_list
+
+        # perform the swap
+        adjacency[u, v] = 0
+        adjacency[v, u] = 0
+        adjacency[x, y] = 0
+        adjacency[y, x] = 0
+
+        adjacency[u, x] = 1
+        adjacency[x, u] = 1
+        adjacency[v, y] = 1
+        adjacency[y, v] = 1
+
+        # DO EDGE LIST STUFF
+        edge_list[orig_inds[0]] = [u, x]
+        edge_list[orig_inds[1]] = [v, y]
```

Comment:
can you remind me if there is a disadvantage to using an array for the edge list?